File: 2021/day1/Day1.hs
import Control.Monad (liftM2, liftM3)

main :: IO ()
main = do
  content <- readFile "input.txt"
  let depths = parseFile content
  putStrLn . show . problem1 $ depths
  putStrLn . show . problem2 $ depths

chunk2 :: [t] -> [(t, t)]
chunk2 = liftM2 zip id (drop 1)

chunk3 :: [t] -> [(t, t, t)]
chunk3 = liftM3 zip3 id (drop 1) (drop 2)

problem1 :: [Int] -> Int
problem1 = length . filter (\(x, y) -> y > x) . chunk2

problem2 :: [Int] -> Int
problem2 = length .
  filter (\(x, y) -> y > x) .
  chunk2 .
  map (\(x, y, z) -> x + y + z) .
  chunk3

parseFile :: String -> [Int]
parseFile = map read . lines

File: 2021/day1/day1.rs
use std::fs::File;
use std::io::prelude::*;
use std::str::FromStr;

fn main() -> () {
    let depths = read_file();

    println!("Problem1: {}", count_increasing_depths(&depths));
    println!(
        "Problem2: {}",
        count_increasing_depths(&combine_depths(&depths))
    );
}

fn count_increasing_depths(depths: &Vec<i32>) -> i32 {
    depths.windows(2).filter(|pair| pair[1] > pair[0]).count() as i32
}

fn combine_depths(depths: &Vec<i32>) -> Vec<i32> {
    depths
        .windows(3)
        .map(|window| window[0] + window[1] + window[2])
        .collect()
}

fn read_file() -> Vec<i32> {
    let mut f = File::open("input.txt").unwrap();
    let mut contents = String::new();

    f.read_to_string(&mut contents).unwrap();

    contents
        .split('\n')
        .filter(|x| x.len() > 0)
        .map(|x| i32::from_str(x).unwrap())
        .collect()
}
